File: matparser/matparser.cpp
#include <stdio.h>
#include <assert.h>
#include <stdint.h>
#include <zlib.h>
#include <vector>
#include <string>

#pragma pack(1)

#define VERBOSE

struct MatHeader
{
    uint8_t signature[116];
    uint64_t offset;
    uint16_t version;
    uint16_t endian;

    void Print() const
    {
        printf("Signature: '%.116s'\n", signature);
        printf("Subsystem offset: 0x%llx\n", offset);
        printf("Version: 0x%x\n", version);
        printf("Endian: %.2s\n", (char*)&endian);
    }
};

enum MatDataType : uint32_t
{
    miINT8      = 1, // 8 bit, signed
    miUINT8     = 2, // 8 bit, unsigned
    miINT16     = 3, // 16-bit, signed
    miUINT16    = 4, // 16-bit, unsigned
    miINT32     = 5, // 32-bit, signed
    miUINT32    = 6, // 32-bit, unsigned
    miSINGLE    = 7, // IEEE 754 single format
    // 8 -- Reserved
    miDOUBLE    = 9, // IEEE 754 double format
    // 10 -- Reserved
    // 11 -- Reserved
    miINT64     = 12, // 64-bit, signed
    miUINT64    = 13, // 64-bit, unsigned
    miMATRIX    = 14, // MATLAB array
    miCOMPRESSED = 15, // Compressed Data
    miUTF8      = 16, // Unicode UTF-8 Encoded Character Data
    miUTF16     = 17, // Unicode UTF-16 Encoded Character Data
    miUTF32     = 18, // Unicode UTF-32 Encoded Character Data
};

size_t MatDataTypeToSize(MatDataType dataType)
{
    switch (dataType)
    {
    case miINT8:
    case miUINT8:
        return 1;

    case miINT16:
    case miUINT16:
        return 2;

    case miINT32:
    case miUINT32:
        return 4;

    case miSINGLE:
        return 4;
    case miDOUBLE:
        return 8;

    case miINT64:
    case miUINT64:
        return 8;

    default:
        return 0;
    }
}

const char* MatDataTypeToString(MatDataType dataType)
{
#define TOSTRING(val) case val: return #val;
    switch (dataType)
    {
    TOSTRING(miINT8)
    TOSTRING(miUINT8)
    TOSTRING(miINT16)
    TOSTRING(miUINT16)
    TOSTRING(miINT32)
    TOSTRING(miUINT32)
    TOSTRING(miSINGLE)
    TOSTRING(miDOUBLE)
    TOSTRING(miINT64)
    TOSTRING(miUINT64)
    TOSTRING(miMATRIX)
    TOSTRING(miCOMPRESSED)
    TOSTRING(miUTF8)
    TOSTRING(miUTF16)
    TOSTRING(miUTF32)    
    default:
        return "Invalid";
    }
#undef TOSTRING
}

struct MatDataElementHeader
{
    MatDataType dataType;
    uint32_t numberOfBytes;

    void Print() const
    {
#ifdef VERBOSE
        printf("\nData Element: %s (%u) size %u\n", MatDataTypeToString(dataType), dataType, numberOfBytes);
#endif
    }
};

enum MatMatrixClass : uint8_t
{
    mxCELL_CLASS    = 1, // Cell array
    mxSTRUCT_CLASS  = 2, // Structure
    mxOBJECT_CLASS  = 3, // Object
    mxCHAR_CLASS    = 4, // Character array
    mxSPARSE_CLASS  = 5, // Sparse array
    mxDOUBLE_CLASS  = 6, // Double precision array
    mxSINGLE_CLASS  = 7, // Single precision array
    mxINT8_CLASS    = 8, // 8-bit, signed integer
    mxUINT8_CLASS   = 9, // 8-bit, unsigned integer
    mxINT16_CLASS   = 10, // 16-bit, signed integer
    mxUINT16_CLASS  = 11, // 16-bit, unsigned integer
    mxINT32_CLASS   = 12, // 32-bit, signed integer
    mxUINT32_CLASS  = 13, // 32-bit, unsigned integer
    mxINT64_CLASS   = 14, // 64-bit, signed integer
    mxUINT64_CLASS  = 15 // 64-bit, unsigned integer
};

const char* MatMatrixClassToString(MatMatrixClass matrixClass)
{
#define TOSTRING(val) case val: return #val;
    switch (matrixClass)
    {
    TOSTRING(mxCELL_CLASS)
    TOSTRING(mxSTRUCT_CLASS)
    TOSTRING(mxOBJECT_CLASS)
    TOSTRING(mxCHAR_CLASS)
    TOSTRING(mxSPARSE_CLASS)
    TOSTRING(mxDOUBLE_CLASS)
    TOSTRING(mxSINGLE_CLASS)
    TOSTRING(mxINT8_CLASS)
    TOSTRING(mxUINT8_CLASS)
    TOSTRING(mxINT16_CLASS)
    TOSTRING(mxUINT16_CLASS)
    TOSTRING(mxINT32_CLASS)
    TOSTRING(mxUINT32_CLASS)
    TOSTRING(mxINT64_CLASS)
    TOSTRING(mxUINT64_CLASS)
    default:
        return "Invalid";
    }
#undef TOSTRING
}

struct MatArrayFlagsSubelement
{
    MatMatrixClass matrixClass;
    union {
        struct {
            unsigned undefined : 4;
            unsigned complex : 1;
            unsigned global : 1;
            unsigned logical : 1;
        };
        uint8_t packed;
    } flags;
    uint16_t undefined;
    uint32_t maxNonZero;

    void Print() const
    {
#ifdef VERBOSE
        printf("ArrayFlags: %s (%u)\n", MatMatrixClassToString(matrixClass), matrixClass);
#endif
    }
};

class MatFile
{
private:
    std::vector<uint8_t> data;
    size_t cursor;

public:
    MatFile(const char* path);
    virtual ~MatFile();

    bool EndOfFile() const;
    void Skip(size_t numberOfBytes);
    MatDataElementHeader ReadDataElementHeader();

    template <class T>
    T ReadPrimitive()
    {
        const T& val = *(const T*)&data[cursor];
        cursor += sizeof(T);
        assert(cursor <= data.size());
        return val;
    }

    template <class T>
    T ReadBlock()
    {
        T block = ReadPrimitive<T>();
        block.Print();
        return block;
    }

    void ReadMem(char* buffer, size_t count)
    {
        memcpy(buffer, &data[cursor], count);
        cursor += count;
        assert(cursor <= data.size());
    }

    void Save(const char* path) const;
};

MatFile::MatFile(const char* path) : cursor(0)
{
    printf("Reading file: %s\n", path);
    FILE* fp = fopen(path, "r+b");
    assert(fp != nullptr);

    fseek(fp, 0, SEEK_END);
    long int fileSize = ftell(fp);
    fseek(fp, 0, SEEK_SET);
    printf("File size: %lu\n", fileSize);
    data.resize(fileSize);
    fread(data.data(), fileSize, 1, fp);
    fclose(fp);

    ReadBlock<MatHeader>();
}

MatFile::~MatFile()
{
}

void MatFile::Save(const char* path) const
{
    printf("Writing file: %s\n", path);
    FILE* fp = fopen(path, "w+b");
    assert(fp != nullptr);

    printf("File size: %lu\n", data.size());
    fwrite(data.data(), data.size(), 1, fp);
    fclose(fp);
}

bool MatFile::EndOfFile() const
{
    assert(cursor <= data.size());
    return cursor == data.size();
}

void MatFile::Skip(size_t numberOfBytes)
{
    assert(cursor + numberOfBytes <= data.size());
    cursor += numberOfBytes;
}

MatDataElementHeader MatFile::ReadDataElementHeader()
{
    MatDataElementHeader header;
    for (;;)
    {
        cursor = ((cursor + 7) & ~7);
        assert((cursor % 8) == 0);
        if (((uint16_t*)&data[cursor])[1] != 0)
        {
            header.dataType = (MatDataType)ReadPrimitive<uint16_t>();
            header.numberOfBytes = ReadPrimitive<uint16_t>();
        }
        else
        {
            header.dataType = ReadPrimitive<MatDataType>();
            header.numberOfBytes = ReadPrimitive<uint32_t>();
        }
        header.Print();

        if (header.dataType == miCOMPRESSED)
        {
            uLongf uncompressedLen = header.numberOfBytes * 20;
            std::vector<uint8_t> uncompressed;
            uncompressed.resize(uncompressedLen);
            int ret = uncompress(uncompressed.data(), &uncompressedLen, (Bytef*)&data[cursor], header.numberOfBytes);
    #define LOG_Z_ERROR(err) if (ret == err) printf("Uncompress returned %s\n", #err)
            LOG_Z_ERROR(Z_MEM_ERROR);
            LOG_Z_ERROR(Z_BUF_ERROR);
            LOG_Z_ERROR(Z_DATA_ERROR);
    #undef LOG_Z_ERROR
            assert(ret == Z_OK);
            uncompressed.resize(uncompressedLen);

            cursor -= sizeof(MatDataElementHeader);
            size_t compressedLen = sizeof(MatDataElementHeader) + header.numberOfBytes;
            size_t remainingLen = data.size() - cursor - compressedLen;
            assert(uncompressedLen > compressedLen);
            data.resize(data.size() - compressedLen + uncompressedLen);
            memcpy(&data[cursor + uncompressedLen], &data[cursor + compressedLen], remainingLen);
            memcpy(&data[cursor], uncompressed.data(), uncompressedLen);
        }
        else
        {
            break;
        }
    }
    return header;
}

size_t g_currentField = (size_t)-1;
std::string g_arrayName;
std::vector<std::string> g_fieldNames;

void ReadMatrixStruct(MatFile& mat)
{
    MatDataElementHeader fieldNameLengthHeader = mat.ReadDataElementHeader();
    assert(fieldNameLengthHeader.dataType == miINT32);
    assert(fieldNameLengthHeader.numberOfBytes == sizeof(int32_t));
    int32_t fieldNameLength = mat.ReadPrimitive<int32_t>();
    printf("\n");

    MatDataElementHeader fieldNamesHeader = mat.ReadDataElementHeader();
    assert(fieldNamesHeader.dataType == miINT8);
    assert(fieldNamesHeader.numberOfBytes % fieldNameLength == 0);
    int32_t fieldNameCount = fieldNamesHeader.numberOfBytes / fieldNameLength;

    std::string fieldName;
    fieldName.resize(fieldNameLength);
    for (int32_t i = 0; i < fieldNameCount; i++)
    {
        mat.ReadMem(&fieldName[0], fieldName.size());
        printf("[%02d]: '%s'\n", i, fieldName.c_str());
        g_fieldNames.push_back(fieldName);
    }
    g_currentField = 0;
}

void ReadMatrixChar(MatFile& mat)
{
    MatDataElementHeader header = mat.ReadDataElementHeader();
    assert(header.dataType == miUTF8);

    std::string charName;
    charName.resize(header.numberOfBytes);
    mat.ReadMem(&charName[0], charName.size());
    printf("%s '%s'\n", MatDataTypeToString(header.dataType), charName.c_str());
}

void ReadMatrixDouble(MatFile& mat)
{
    MatDataElementHeader header = mat.ReadDataElementHeader();
    size_t dataTypeSize = MatDataTypeToSize(header.dataType);
    assert(dataTypeSize != 0);
    printf("%s\n", MatDataTypeToString(header.dataType));
    mat.Skip(header.numberOfBytes);
}

void ReadMatrix(MatFile& mat)
{
    MatDataElementHeader arrayFlagsHeader = mat.ReadDataElementHeader();
    assert(arrayFlagsHeader.dataType == miUINT32);
    assert(arrayFlagsHeader.numberOfBytes == 2 * sizeof(uint32_t));
    MatArrayFlagsSubelement arrayFlags = mat.ReadBlock<MatArrayFlagsSubelement>();

    printf("\n");
    if (!g_fieldNames.empty())
    {
        assert(g_currentField < g_fieldNames.size());
        printf("%s.%s: ", g_arrayName.c_str(), g_fieldNames[g_currentField].c_str());
        g_currentField++;
    }

    {
        MatDataElementHeader dimensionHeader = mat.ReadDataElementHeader();
        assert(dimensionHeader.dataType == miINT32);
        uint32_t dimensions = dimensionHeader.numberOfBytes / sizeof(uint32_t);
        for (uint32_t i = 0; i < dimensions; i++)
        {
            printf("%d", mat.ReadPrimitive<int32_t>());
            if (i < dimensions - 1)
                printf("-by-");
        }
        printf(", ");
    }

    {
        MatDataElementHeader arrayNameHeader = mat.ReadDataElementHeader();
        assert(arrayNameHeader.dataType == miINT8);
        if (arrayNameHeader.numberOfBytes > 0)
        {
            g_arrayName.resize(arrayNameHeader.numberOfBytes);
            mat.ReadMem(&g_arrayName[0], g_arrayName.size());
            printf("'%s' ", g_arrayName.c_str());
        }
    }

    switch (arrayFlags.matrixClass)
    {
        case mxSTRUCT_CLASS:
            ReadMatrixStruct(mat);
            break;

        case mxCHAR_CLASS:
            ReadMatrixChar(mat);
            break;

        case mxDOUBLE_CLASS:
            ReadMatrixDouble(mat);
            break;

        default:
            printf("Unsupported matrix type!\n");
            assert(0);
    }
}

int main()
{
    //MatFile mat("fctd_grid_uncompressed.mat");
    MatFile mat("epsi_grid_uncompressed.mat");

    while (!mat.EndOfFile())
    {
        MatDataElementHeader element = mat.ReadDataElementHeader();
        if (element.dataType == miMATRIX)
        {
            ReadMatrix(mat);
        }
        else
        {
            mat.Skip(element.numberOfBytes);
        }
    }
    //mat.Save("fctd_test.mat");
    return 0;
}

File: ChartTest1/ChartTest1/MatParser.swift
import Foundation
import Compression

let VERBOSE = false

class MatHeader {
    var signature = "" // 116
    var offset = 0
    var version = 0
    var endian = "" // 2

    public func print() {
        Swift.print("Signature: '\(signature)'");
        Swift.print("Subsystem offset: 0x\(String(format:"%x", offset))")
        Swift.print("Version: 0x\(String(format:"%02X", version))")
        Swift.print("Endian: '\(endian)'")
    }
}

enum MatDataType : Int {
    case miINT8      = 1 // 8 bit, signed
    case miUINT8     = 2 // 8 bit, unsigned
    case miINT16     = 3 // 16-bit, signed
    case miUINT16    = 4 // 16-bit, unsigned
    case miINT32     = 5 // 32-bit, signed
    case miUINT32    = 6 // 32-bit, unsigned
    case miSINGLE    = 7 // IEEE 754 single format
    // 8 -- Reserved
    case miDOUBLE    = 9 // IEEE 754 double format
    // 10 -- Reserved
    // 11 -- Reserved
    case miINT64     = 12 // 64-bit, signed
    case miUINT64    = 13 // 64-bit, unsigned
    case miMATRIX    = 14 // MATLAB array
    case miCOMPRESSED = 15 // Compressed Data
    case miUTF8      = 16 // Unicode UTF-8 Encoded Character Data
    case miUTF16     = 17 // Unicode UTF-16 Encoded Character Data
    case miUTF32     = 18 // Unicode UTF-32 Encoded Character Data
}

func MatDataTypeToSize(_ dataType: MatDataType) -> Int {
    switch (dataType)
    {
    case MatDataType.miINT8:   return 1
    case MatDataType.miUINT8:  return 1
    case MatDataType.miINT16:  return 2
    case MatDataType.miUINT16: return 2

    case MatDataType.miINT32:  return 4
    case MatDataType.miUINT32: return 4

    case MatDataType.miSINGLE: return 4
    case MatDataType.miDOUBLE: return 8

    case MatDataType.miINT64:  return 8
    case MatDataType.miUINT64: return 8

    default:
        return 0
    }
}

class MatDataElementHeader {
    var dataType = MatDataType.miINT8
    var numberOfBytes = 0

    func print(force : Bool = false) {
        if (VERBOSE || force) {
            Swift.print("\nData Element: \(String(describing: dataType)) (\(dataType.rawValue)) size \(numberOfBytes)")
        }
    }
}

enum MatMatrixClass : Int {
    case mxCELL_CLASS    = 1 // Cell array
    case mxSTRUCT_CLASS  = 2 // Structure
    case mxOBJECT_CLASS  = 3 // Object
    case mxCHAR_CLASS    = 4 // Character array
    case mxSPARSE_CLASS  = 5 // Sparse array
    case mxDOUBLE_CLASS  = 6 // Double precision array
    case mxSINGLE_CLASS  = 7 // Single precision array
    case mxINT8_CLASS    = 8 // 8-bit, signed integer
    case mxUINT8_CLASS   = 9 // 8-bit, unsigned integer
    case mxINT16_CLASS   = 10 // 16-bit, signed integer
    case mxUINT16_CLASS  = 11 // 16-bit, unsigned integer
    case mxINT32_CLASS   = 12 // 32-bit, signed integer
    case mxUINT32_CLASS  = 13 // 32-bit, unsigned integer
    case mxINT64_CLASS   = 14 // 64-bit, signed integer
    case mxUINT64_CLASS  = 15 // 64-bit, unsigned integer
}

class MatArrayFlagsSubelement {
    var matrixClass = MatMatrixClass.mxCELL_CLASS

    func print(force : Bool = false) {
        if (VERBOSE || force) {
            Swift.print("ArrayFlags: \(String(describing: matrixClass)) (\(matrixClass))")
        }
    }
}

class MatFile {
    private var data : [UInt8]
    private var cursor : Int
    private var cursor0 : Int

    public init(_ path : String) {
        print("Reading file: \(path)")

        let fileData = try! Data(contentsOf: URL(fileURLWithPath: path))
        self.data = [UInt8](repeating: 0, count: fileData.count)
        fileData.copyBytes(to: &self.data, count: self.data.count)

        cursor = 0
        cursor0 = 0
        _ = readMatHeader()
        cursor0 = cursor

        // Do we need to decompress the file?
        let firstDataElement = readDataElementHeader()
        if (firstDataElement.dataType == MatDataType.miCOMPRESSED) {
            // We only support a single compressed chunk
            assert(data.count == 8 + firstDataElement.numberOfBytes)
            let uncompressedLen = firstDataElement.numberOfBytes * 20;
            let uncompressed = UnsafeMutablePointer<UInt8>.allocate(capacity: uncompressedLen)
            // 2 is a mysterious fudge factor
            let writtenLen = data[(cursor + 2)..<data.count].withUnsafeBytes ({
                return compression_decode_buffer(
                    uncompressed, uncompressedLen,
                    $0.baseAddress!.bindMemory(to: UInt8.self, capacity: 1),
                    firstDataElement.numberOfBytes - 2, nil, COMPRESSION_ZLIB)
            })
            assert(writtenLen > 0)
            // Replace original data with decompressed so we can continue parsing
            data = Array(UnsafeBufferPointer(start: uncompressed, count: writtenLen))
            cursor0 = 0
            uncompressed.deallocate()
        }
        // Rewind, whether we just decompressed or not
        seek(0)
    }

    public func endOfFile() -> Bool {
        assert(cursor <= data.count)
        return cursor == data.count
    }

    public func skip(_ numberOfBytes : Int) {
        assert(cursor + numberOfBytes <= data.count)
        cursor += numberOfBytes;
    }

    public func seek(_ pos : Int) {
        assert(cursor0 + pos >= 0 && cursor0 + pos < data.count)
        cursor = cursor0 + pos;
    }

    public func tell() -> Int {
        assert(cursor >= cursor0)
        return cursor - cursor0
    }

    public func readByte() -> UInt8 {
        assert(cursor < data.count)
        defer { cursor += 1 }
        return data[cursor]
    }

    public func readChar() -> Character {
        let byte = readByte()
        return Character(UnicodeScalar(byte))
    }

    public func readString(_ numChars: Int, _ encoding: MatDataType? = nil) -> String {
        var str = ""
        for _ in 0..<numChars {
            let byte = readByte()
            if (byte != 0) {
                str += String(Character(UnicodeScalar(byte)))
            }
        }

        if (encoding != nil) {
            switch (encoding!) {
            case .miUTF8:
                str = String(str.utf8)

            case .miUTF16:
                str = String(str.utf16)

            default:
                print("ERROR: Unsupported string encoding: \(String(describing: encoding!))")
                assert(false)
            }
        }

        return str
    }

    public func readLEDouble() -> Double
    {
        let expected = MemoryLayout<Double>.size
        assert(cursor + expected <= data.count)
        defer { cursor += expected }
        return data.withUnsafeBytes {
            return $0.load(fromByteOffset: cursor, as: Double.self)
        }
    }

    public func readLEUIntX<Result>(_: Result.Type) -> Result
            where Result: UnsignedInteger
    {
        let expected = MemoryLayout<Result>.size
        assert(cursor + expected <= data.count)
        defer { cursor += expected }
        let sub = data[cursor..<cursor + expected]
        return sub
            .reversed()
            .reduce(0, { soFar, new in
                    (soFar << 8) | Result(new)
            })
    }

    public func readLEUInt8() -> Int {
        Int(readLEUIntX(UInt8.self))
    }

    public func readLEUInt16() -> Int {
        Int(readLEUIntX(UInt16.self))
    }

    public func readLEUInt32() -> Int {
        Int(readLEUIntX(UInt32.self))
    }

    public func readLEUInt64() -> Int {
        Int(readLEUIntX(UInt64.self))
    }

    public func readMatHeader() -> MatHeader {
        assert(cursor == 0)
        let header = MatHeader()
        header.signature = readString(116)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        header.offset = readLEUInt64()
        header.version = readLEUInt16()
        header.endian = readString(2)
        assert(header.version == 0x100)
        assert(header.endian == "IM")
        header.print()
        return header
    }

    public func readDataElementHeader() -> MatDataElementHeader {
        let header = MatDataElementHeader()
        // Align start of data element to the next 64bit boundary
        cursor = ((cursor + 7) & ~7)
        // Is it a small data element, packed into 4bytes?
        if (data[cursor + 2] != 0 || data[cursor + 3] != 0)
        {
            header.dataType = MatDataType(rawValue: readLEUInt16())!
            header.numberOfBytes = readLEUInt16()
        }
        else // No, regular sized 8byte data element
        {
            header.dataType = MatDataType(rawValue: readLEUInt32())!
            header.numberOfBytes = readLEUInt32()
        }
        header.print()
        return header
    }

    public func readArrayFlags() -> MatArrayFlagsSubelement {
        let arrayFlags = MatArrayFlagsSubelement()
        arrayFlags.matrixClass = MatMatrixClass(rawValue: readLEUInt8())!
        skip(7)
        arrayFlags.print()
        return arrayFlags
    }
}

class MatData {
    struct FieldInfo {
        var name = ""
        var offsetInFile = 0
        var dimensions : [Int] = [Int]()
    }

    var mat : MatFile
    var arrayName = ""
    var fieldInfo = [FieldInfo]()

    init(path : String) {
        mat = MatFile(path)
        while !mat.endOfFile() {
            let element = mat.readDataElementHeader();
            switch (element.dataType) {
            case MatDataType.miMATRIX:
                readMatrix()

            default:
                print("ERROR: Unknown data element!")
                element.print(force: true)
                assert(false)
                //mat.skip(element.numberOfBytes)
            }
        }
        mat.seek(0)
    }

    private func readMatrixStruct() {
        let fieldNameLengthHeader = mat.readDataElementHeader()
        assert(fieldNameLengthHeader.dataType == MatDataType.miINT32)
        assert(fieldNameLengthHeader.numberOfBytes == MemoryLayout<Int32>.size)
        let fieldNameLength = mat.readLEUInt32()
        print()
        
        let fieldNamesHeader = mat.readDataElementHeader();
        assert(fieldNamesHeader.dataType == MatDataType.miINT8)
        assert(fieldNamesHeader.numberOfBytes % fieldNameLength == 0);
        let fieldNameCount = fieldNamesHeader.numberOfBytes / fieldNameLength
        
        fieldInfo = [FieldInfo](repeating: FieldInfo(), count: fieldNameCount)
        for i in 0..<fieldNameCount {
            let fieldName = mat.readString(fieldNameLength)
            print("[\(String(format: "%02d", i))]: '\(fieldName)'")
            fieldInfo[i].name = fieldName
        }
        for i in 0..<fieldNameCount {
            let field = mat.readDataElementHeader();
            assert(field.dataType == MatDataType.miMATRIX)
            readMatrix(currentField: i)
        }
    }
    
    func skipMatrixChar() {
        let header = mat.readDataElementHeader()
        let val = mat.readString(header.numberOfBytes, header.dataType)
        print("\(String(describing: header.dataType)) '\(val)'")
    }
    
    func getMatrixChar(name: String) -> String {
        let field = fieldInfo.first { $0.name == name }
        assert(field != nil)

        let prevPos = mat.tell()
        defer { mat.seek(prevPos) }
        mat.seek(field!.offsetInFile)

        let header = mat.readDataElementHeader()
        return mat.readString(header.numberOfBytes, header.dataType)
    }

    func skipMatrixDouble() {
        let header = mat.readDataElementHeader()
        let dataTypeSize = MatDataTypeToSize(header.dataType)
        assert(dataTypeSize != 0)
        print(String(describing: header.dataType))
        if (header.dataType == MatDataType.miDOUBLE) {
            var minVal = mat.readLEDouble()
            var maxVal = minVal
            for _ in 1..<(header.numberOfBytes / dataTypeSize) {
                let val = mat.readLEDouble()
                minVal = fmin(minVal, val)
                maxVal = fmax(maxVal, val)
            }
            print("Min: \(minVal), Max: \(maxVal)")
        } else {
            mat.skip(header.numberOfBytes)
        }
        //mat.skip(header.numberOfBytes)
    }
    
    func getMatrixDouble2(name : String) -> [[Double]] {
        let field = fieldInfo.first { $0.name == name }
        assert(field != nil)

        let prevPos = mat.tell()
        defer { mat.seek(prevPos) }
        mat.seek(field!.offsetInFile)

        assert(field!.dimensions.count == 2)
        var result = Array(repeating: Array<Double>(repeating: 0.0, count: field!.dimensions[1]), count: field!.dimensions[0])

        let header = mat.readDataElementHeader()
        for i in 0..<result[0].count {
            for j in 0..<result.count {
                switch (header.dataType) {
                case .miDOUBLE:
                    result[j][i] = mat.readLEDouble()

                case .miUINT8:
                    result[j][i] = Double(mat.readLEUInt8())

                case .miUINT16:
                    result[j][i] = Double(mat.readLEUInt16())

                default:
                    print("Unsupported Matrix2D type: \(String(describing: header.dataType))")
                    assert(false)
                }
            }
        }

        return result
    }

    func getFieldNames(filter2D : Bool) -> [String] {
        var fieldNames = [String]()
        for i in 0..<fieldInfo.count {
            let field = fieldInfo[i]
            if (filter2D) {
                if !(field.dimensions.count == 2 &&
                    field.dimensions[0] != 1 &&
                    field.dimensions[1] != 1) {
                    continue
                }
            }
            fieldNames.append("\(arrayName).\(field.name)")
        }
        return fieldNames
    }

    func getArrayName() -> String {
        return arrayName
    }

    func readMatrix(currentField: Int? = nil) {
        let arrayFlagsHeader = mat.readDataElementHeader()
        assert(arrayFlagsHeader.dataType == MatDataType.miUINT32)
        assert(arrayFlagsHeader.numberOfBytes == 2 * MemoryLayout<UInt32>.size)
        let arrayFlags = mat.readArrayFlags()
        
        print()
        if (currentField != nil) {
            print("\(arrayName).\(fieldInfo[currentField!].name): ", terminator: "")
        }
        
        let dimensionHeader = mat.readDataElementHeader()
        assert(dimensionHeader.dataType == MatDataType.miINT32)
        let dimensions = dimensionHeader.numberOfBytes / MemoryLayout<UInt32>.size
        for i in 0..<dimensions {
            let val = mat.readLEUInt32()
            if (currentField != nil) {
                fieldInfo[currentField!].dimensions.append(val)
            }
            print(val, terminator: "")
            if (i < dimensions - 1) {
                print("-by-", terminator: "")
            }
        }
        print(", ", terminator: "")
        
        let arrayNameHeader = mat.readDataElementHeader()
        assert(arrayNameHeader.dataType == MatDataType.miINT8)
        if (arrayNameHeader.numberOfBytes > 0) {
            arrayName = mat.readString(arrayNameHeader.numberOfBytes)
            print("'\(arrayName)' ", terminator: "")
        }

        if (currentField != nil) {
            fieldInfo[currentField!].offsetInFile = mat.tell()
        }

        switch (arrayFlags.matrixClass) {
        case MatMatrixClass.mxSTRUCT_CLASS:
            assert(currentField == nil)
            readMatrixStruct()

        case MatMatrixClass.mxCHAR_CLASS:
            assert(currentField != nil)
            skipMatrixChar()
            
        case MatMatrixClass.mxDOUBLE_CLASS:
            assert(currentField != nil)
            skipMatrixDouble()
            
        default:
            print("ERROR: Unsupported matrix type!")
            arrayFlags.print(force: true)
            assert(false)
        }
    }
}
